### ISSUE TITLE
Problems with whole program analysis and Java 8?

### ISSUE THREAD

**Issue by ChuckFoo** [Sable/soot#874]
While investigating some Soot call graph issues in some analysis code I'm working on, I updated to the latest & greatest Soot to pick up bugfixes, which resulted in me having to switch from using Java 7 (which I had been using previously after hitting issues w/ Java 8 which seemed to be known to the developers at the time) to Java 8 due to recent commits.  I assumed that this meant that the problems with using Java 8 to run the analysis had gone away, but apparently some still exist. :)

The first problem can be seen by doing whole program analysis on something like a recent JEdit installer (not what I was working with but a simple test case that shows the same issue w/ Soot standalone, jar with deps downloaded from Soot nightly builds 1/16/18), like so:

```
java -classpath ~/Downloads/sootclasses-trunk-jar-with-dependencies.jar soot.Main -w -p cg verbose:true -p cg.spark "enabled:true,verbose:true,pre-jimplify:true" -verbose -debug -soot-class-path ~/Downloads/jedit5.4.0install.jar:/usr/lib/jvm/java-1.8.0-openjdk-amd64/jre/lib/rt.jar:/usr/lib/jvm/java-1.8.0-openjdk-amd64/jre/lib/jce.jar -process-dir ~/Downloads/jedit5.4.0install.jar
```

This results in the following exception:

```
java.lang.RuntimeException: Failed to convert <java.util.stream.FindOps: java.util.stream.TerminalOp makeRef(boolean)>
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:1883)
	at soot.SootMethod.getBodyFromMethodSource(SootMethod.java:126)
	at soot.SootMethod.retrieveActiveBody(SootMethod.java:385)
	at soot.jimple.spark.builder.ContextInsensitiveBuilder.preJimplify(ContextInsensitiveBuilder.java:58)
	at soot.jimple.spark.SparkTransformer.internalTransform(SparkTransformer.java:80)
	at soot.SceneTransformer.transform(SceneTransformer.java:39)
	at soot.Transform.apply(Transform.java:90)
	at soot.RadioScenePack.internalApply(RadioScenePack.java:57)
	at soot.jimple.toolkits.callgraph.CallGraphPack.internalApply(CallGraphPack.java:49)
	at soot.Pack.apply(Pack.java:118)
	at soot.PackManager.runWholeProgramPacks(PackManager.java:564)
	at soot.PackManager.runPacksNormally(PackManager.java:455)
	at soot.PackManager.runPacks(PackManager.java:384)
	at soot.Main.run(Main.java:278)
	at soot.Main.main(Main.java:147)
Caused by: java.lang.RuntimeException: 
Aborting: can't find classfile java.util.stream.FindOps$FindSink$OfRef
	at soot.Scene.getSootClass(Scene.java:991)
	at soot.asm.AsmMethodSource.toSootMethodRef(AsmMethodSource.java:1327)
	at soot.asm.AsmMethodSource.toSootValue(AsmMethodSource.java:1107)
	at soot.asm.AsmMethodSource.convertInvokeDynamicInsn(AsmMethodSource.java:1250)
	at soot.asm.AsmMethodSource.convert(AsmMethodSource.java:1639)
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:1881)
	... 14 more
```

And that "missing" classfile certainly appears to be in the same rt.jar that java.util.stream.FindOps is in:

```
$ jar tvf /usr/lib/jvm/java-1.8.0-openjdk-amd64/jre/lib/rt.jar | grep 'FindOps$FindSink$OfRef'
   815 Fri Oct 27 22:02:06 EDT 2017 java/util/stream/FindOps$FindSink$OfRef.class
```

So the fact that it can't find it is pretty confusing to me.

Now, a second slightly different problem (perhaps deserving of being a separate issue) happens if I remove the pre-jimplify:true setting from the above command line:

```
WARNING: InvokeDynamic to dynamicinvoke "run" <java.security.PrivilegedAction (int,int,int,int,int,java.awt.Image,java.awt.image.ImageObserver)>(i4, i3, i2, i1, i2, r9, r0) <java.lang.invoke.LambdaMetafactory: java.lang.invoke.CallSite metafactory(java.lang.invoke.MethodHandles$Lookup,java.lang.String,java.lang.invoke.MethodType,java.lang.invoke.MethodType,java.lang.invoke.MethodHandle,java.lang.invoke.MethodType)>(class "()Ljava/lang/Object;", handle: <sun.awt.image.ImageWatched$WeakLink: java.lang.Boolean lambda$update$0(java.awt.image.ImageObserver,java.awt.Image,int,int,int,int,int)>, class "()Ljava/lang/Boolean;") not resolved during call-graph construction.
java.lang.RuntimeException: Attempt to create VarNode of type int
	at soot.jimple.spark.pag.VarNode.<init>(VarNode.java:106)
	at soot.jimple.spark.pag.LocalVarNode.<init>(LocalVarNode.java:44)
	at soot.jimple.spark.pag.PAG.makeLocalVarNode(PAG.java:711)
	at soot.jimple.spark.builder.MethodNodeFactory.caseLocal(MethodNodeFactory.java:319)
	at soot.jimple.internal.JimpleLocal.apply(JimpleLocal.java:125)
	at soot.jimple.spark.builder.MethodNodeFactory.caseCastExpr(MethodNodeFactory.java:296)
	at soot.jimple.internal.AbstractCastExpr.apply(AbstractCastExpr.java:134)
	at soot.jimple.spark.builder.MethodNodeFactory$1.caseAssignStmt(MethodNodeFactory.java:139)
	at soot.jimple.internal.JAssignStmt.apply(JAssignStmt.java:238)
	at soot.jimple.spark.builder.MethodNodeFactory.handleStmt(MethodNodeFactory.java:128)
	at soot.jimple.spark.pag.MethodPAG.buildNormal(MethodPAG.java:181)
	at soot.jimple.spark.pag.MethodPAG.build(MethodPAG.java:150)
	at soot.jimple.spark.solver.OnFlyCallGraph.processReachables(OnFlyCallGraph.java:83)
	at soot.jimple.spark.solver.OnFlyCallGraph.build(OnFlyCallGraph.java:75)
	at soot.jimple.spark.solver.PropWorklist.handleVarNode(PropWorklist.java:152)
	at soot.jimple.spark.solver.PropWorklist.propagate(PropWorklist.java:75)
	at soot.jimple.spark.SparkTransformer.internalTransform(SparkTransformer.java:153)
	at soot.SceneTransformer.transform(SceneTransformer.java:39)
	at soot.Transform.apply(Transform.java:90)
	at soot.RadioScenePack.internalApply(RadioScenePack.java:57)
	at soot.jimple.toolkits.callgraph.CallGraphPack.internalApply(CallGraphPack.java:49)
	at soot.Pack.apply(Pack.java:118)
	at soot.PackManager.runWholeProgramPacks(PackManager.java:564)
	at soot.PackManager.runPacksNormally(PackManager.java:455)
	at soot.PackManager.runPacks(PackManager.java:384)
	at soot.Main.run(Main.java:278)
	at soot.Main.main(Main.java:147)
```

Any ideas what might be going on with both of these problems and/or hints on how to work around the problems temporarily?

**Comment by ChuckFoo** [comment]
Oh, forgot to mention that this was with stock OpenJDK 8u151 under Ubuntu 16.04 but also happened with 8u152 downloaded from Oracle.

**Comment by patricklam** [comment]
Have you tried running Soot under Java 8 with a soot classpath from Java 7?

**Comment by ChuckFoo** [comment]
You know, I'm a bit disappointed that didn't occur to me to try as a temporary workaround. :(

Yeah, that did indeed seem to get me past these particular issues so I can get back to debugging the original issues I was investigating at least.  Thanks.

**Comment by mbenz89** [comment]
Duplicate #875

**Comment by anddann** [comment]
Thank you for reporting this issue.

**Comment by anddann** [comment]
After a quick investigation, I discovered the following problems:
- the inner classes java.util.stream.FindOps$... are not properly resolved
- the order of arguments when constructing dynamic invoke expressions is wrong
I created a dirty fix in my forked repository https://github.com/anddann/soot/tree/issue874
Applying this fix works on the jedit_installer jar with the arguments you provided.
However, before doing a pull request I want to clean up the resolving (especially the resolving of the inner classes). IMHO: It currently looks like a dirty workaround.

**Comment by ryanphuang** [comment]
I'm experiencing the same problem with the same exception stacktrace. Using JDK 1.7's rt.jar in the classpath does not resolve the problem for me. @anddann, thank you for the fix, will your plan to submit your fix to the main repo? I'm using the soot snapshot from the nexus repo of Maven. It'd be great if the fix will be merged there.

```
soot.SootResolver$SootClassNotFoundException: couldn't find class: javax.crypto.spec.DESKeySpec (is your soot-class-path set properly?)
        at soot.SootResolver.bringToHierarchyUnchecked(SootResolver.java:228)
        at soot.SootResolver.bringToHierarchy(SootResolver.java:208)
        at soot.SootResolver.bringToSignatures(SootResolver.java:267)
        at soot.SootResolver.bringToBodies(SootResolver.java:312)
        at soot.SootResolver.processResolveWorklist(SootResolver.java:163)
        at soot.SootResolver.resolveClass(SootResolver.java:133)
        at soot.Scene.tryLoadClass(Scene.java:830)
        at soot.Scene.loadBasicClasses(Scene.java:1503)
        at soot.Scene.loadNecessaryClasses(Scene.java:1586)
        at soot.Main.run(Main.java:250)
        at soot.Main.main(Main.java:147)
        at edu.jhu.order.observatory.Main.main(Main.java:55)
```

**Comment by anddann** [comment]
@ryanphuang the stacktrace you posted looks quite different.
Anyway, did my fix solve your problem?
For testing  you can find a jar  here https://soot-build.cs.upb.de/public/origin/java9/soot/soot-j9/
and the maven dependency is 
```
<dependency>
    <groupId>ca.mcgill.sable</groupId>
    <artifactId>soot</artifactId>
    <version>3.0.0-SNAPSHOT</version>
  </dependency>
</dependencies>
 ```
Yes, I plan to do a pull request and review the "fix" with @mbenz89. 
Thus, it would be helpful if you let us know, if the fix solved your problem

**Comment by ChuckFoo** [comment]
@anddann I assume the question was really for @ryanphuang and not me, but just in case, I did pull your "dirty hack" fixes and they do seem to fix both both problems I reported for the jedit installer example and for another thing I just tried real quick as well.

**Comment by ryanphuang** [comment]
@anddann yes the j9 branch build fixes my problem! thanks a lot!

**Comment by anddann** [comment]
@ChuckFoo @ryanphuang thank you both for testing. Due to your fast feedback, I was able to create a pull request. @mbenz89 will review it and hopefully merge it.

**Comment by mbenz89** [comment]
Fixed in #882
